mlflow/telemetry/client.py

```diff
@@ -162,15 +178,28 @@ def _process_records(self, records: list[APIRecord]):
                 }
                 for record in records
             ]
-            # TODO: add retry logic
-            response = requests.post(
-                self.config.ingestion_url,
-                json={"records": records},
-                headers={"Content-Type": "application/json"},
-                timeout=3,
-            )
-            if response.status_code != 200:
-                pass
+            max_retries = 3
+            for i in range(max_retries):
+                response = requests.post(
+                    self.config.ingestion_url,
+                    json={"records": records},
+                    headers={"Content-Type": "application/json"},
+                    timeout=timeout,
+                )
+                # if this is executed when terminating, we should not retry
+                if self._is_stopped:
+                    return
+                if response.status_code in STOP_COLLECTION_ERRORS:
+                    self._is_stopped = True
+                    self.is_active = False
+                    # this is executed in the consumer thread, so
+                    # we cannot join the thread here, but this should
+                    # be enough to stop the telemetry collection
+                    return
+                if response.status_code in RETRYABLE_ERRORS:
+                    time.sleep(2**i)
```

Comment:
can we skip sleep in the final attempt?

Reply:
Yes we should, thanks for catching this!